File: sdlf-ddk-lightweight/typescript/src/datalake/pipelines/sdlf-base-stack.ts
import * as path from "path";
import { readFileSync } from "fs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { Construct } from "constructs";
import { BaseStack, BaseStackProps } from "aws-ddk-core";

import { FoundationsStack } from "../foundations";
import { StandardPipeline } from "./standard-pipeline";
import { CustomPipeline } from "./custom-pipeline";



export interface SDLFBaseStackProps extends BaseStackProps {
    environmentId: string;
    resourcePrefix: string;
    params: any;
}

export class SDLFBaseStack extends BaseStack {
    readonly resourcePrefix: string;
    readonly params: any;
    readonly app: string;
    readonly org: string;
    readonly wranglerLayer: lambda.ILayerVersion;
    readonly foundationsStage: FoundationsStack;

    constructor(scope: Construct, id: string, props: SDLFBaseStackProps) {
        super(scope, id, props);
        this.resourcePrefix = props.resourcePrefix;
        this.params = props.params;
        this.app = this.params.app ?? "datalake";
        this.org = this.params.org ?? "aws";
        const customerConfigs: any = JSON.parse(readFileSync("./src/pipelines/parameters.json", "utf-8"));

        this.wranglerLayer = this.createWranglerLayer()
        this.createDatalakeLibraryLayer()

        // creates DDB tables, routing lambda, data lake buckets,
        // pushes Glue scripts to S3, and creates data lake library lamdba layer
        this.foundationsStage = new FoundationsStack(
            this,
            "foundation-stage",
            {
                environmentId: props.environmentId,
                resourcePrefix: this.resourcePrefix,
                app: this.app,
                org: this.org,
                runtime: lambda.Runtime.PYTHON_3_9
            }
        )

        const datasetNames: any = []
        const pipelines: any = {}
        
        // loop through values in parameters.json and create the necessary resources for each pipeline
        customerConfigs.environmentId.forEach( (customerConfig: any) => {
            const dataset = customerConfig.dataset
            const team = customerConfig.team
            const pipelineType = customerConfig.pipeline ?? "standard"

        // PIPELINE CREATION
            const pipelineName = `${team}-${pipelineType}`
            var pipeline: StandardPipeline | CustomPipeline;
            if (pipelineType == "standard") {
                pipeline = new StandardPipeline(
                    this,
                    `${team}-${pipelineType}`,
                    {
                        environmentId: props.environmentId,
                        resourcePrefix: this.resourcePrefix,
                        team: team,
                        foundationsStage: this.foundationsStage,
                        wranglerLayer: this.wranglerLayer,
                        app: this.app,
                        org: this.org,
                        runtime: lambda.Runtime.PYTHON_3_9
                    }
                )
            }
            else if (pipelineType == "custom"){
                pipeline = new CustomPipeline(
                    this,
                    `${team}-${pipelineType}`,
                    {
                        environmentId: props.environmentId,
                        resourcePrefix: this.resourcePrefix,
                        team: team,
                        foundationsStage: this.foundationsStage,
                        wranglerLayer: this.wranglerLayer,
                        app: this.app,
                        org: this.org,
                        runtime: lambda.Runtime.PYTHON_3_9
                    }
                )
            }
            else {
                throw new Error(`Could not find a valid implementation for pipeline type: ${pipelineType}`);
            }
            pipelines[pipelineName] = pipeline


            // Register dataset to pipeline with concrete implementations
            const datasetName = `${team}-${dataset}`;
            if (!datasetName.includes(datasetName)) {
                datasetNames.add(datasetName)
                pipeline.registerDataset(
                    dataset,
                    customerConfig.config ?? {}
                )
            }
        }
    }
    protected createWranglerLayer(): lambda.ILayerVersion {
        return lambda.LayerVersion.fromLayerVersionArn(
            this,
            "wrangler-layer",
            `arn:aws:lambda:${this.region}:336392948345:layer:AWSDataWrangler-Python39:1`,
        )
    }
    protected createDatalakeLibraryLayer(): void {
        const datalakeLibraryLayer = new lambda.LayerVersion(
            this,
            "data-lake-library-layer",
            {
                layerVersionName: "data-lake-library",
                code: lambda.Code.fromAsset(
                    path.join(
                        __dirname, "src/layers/datalakelibrary"
                    )
                ),
                compatibleRuntimes: [lambda.Runtime.PYTHON_3_9],
                description: `${this.resourcePrefix} Data Lake Library`,
                license: "Apache-2.0",
            }
        )

        new ssm.StringParameter(
            this,
            `data-lake-library-layer-ssm`,
            {
                parameterName: "/SDLF/Layer/DataLakeLibrary",
                stringValue: datalakeLibraryLayer.layerVersionArn,
            }
        )

    }
}

File: sdlf-ddk-lightweight/python/data_lake/pipelines/sdlf_base_stack.py
# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import json
import os
from pathlib import Path
from typing import Any, Dict, Protocol
import aws_cdk as cdk
import aws_cdk.aws_iam as iam
import aws_cdk.aws_lambda as lmbda
import aws_cdk.aws_ssm as ssm
from aws_ddk_core import BaseStack
from constructs import Construct

from ..foundations import FoundationsStack
from .custom_pipeline import CustomPipeline
from .standard_pipeline import StandardPipeline


class SDLFPipeline(Protocol):
    PIPELINE_TYPE: str

    def register_dataset(self, dataset: str, config: Dict[str, Any]):
        ...


class SDLFBaseStack(BaseStack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_id: str,
        resource_prefix: str,
        params: Dict[str, Any],
        **kwargs: Any,
    ) -> None:
        self._resource_prefix = resource_prefix
        super().__init__(
            scope,
            construct_id,
            environment_id=environment_id,
            stack_name=f"{self._resource_prefix}-SDLFPipelineStack-{environment_id}",
            **kwargs,
        )
        self._params = params
        self._environment_id = environment_id
        self._app = self._params.get("app", "datalake")
        self._org = self._params.get("org", "aws")

        path = os.path.join(f"{Path(__file__).parents[0]}", "parameters.json")
        with open(path, encoding="utf-8") as params_file:
            customer_configs = json.load(params_file).get(self._environment_id)

        self._wrangler_layer = self._create_wrangler_layer()
        self._create_data_lake_library_layer()

        # creates DDB tables, routing lambda, data lake buckets,
        # pushes Glue scripts to S3, and creates data lake library lamdba layer
        self._foundations_stage = FoundationsStack(
            self,
            construct_id="foundation-stage",
            environment_id=self._environment_id,
            resource_prefix=self._resource_prefix,
            app=self._app,
            org=self._org,
            runtime=lmbda.Runtime.PYTHON_3_9,
        )

        dataset_names: set[str] = set()
        pipelines: Dict[str, SDLFPipeline] = {}
        
        # loop through values in parameters.json and create the necessary resources for each pipeline
        for customer_config in customer_configs:
            dataset = customer_config["dataset"]
            team = customer_config["team"]
            pipeline_type = customer_config.get(
                "pipeline", StandardPipeline.PIPELINE_TYPE
            )
            orchestration = customer_config.get(
                "orchestration", "sfn"
            ).lower()

            # PIPELINE CREATION
            pipeline: SDLFPipeline
            pipeline_name = f"{team}-{pipeline_type}"
            if pipeline_name not in pipelines:
                if pipeline_type == StandardPipeline.PIPELINE_TYPE:
                    pipeline = StandardPipeline(
                        self,
                        construct_id=f"{team}-{pipeline_type}",
                        environment_id=self._environment_id,
                        resource_prefix=self._resource_prefix,
                        team=team,
                        orchestration=orchestration,
                        foundations_stage=self._foundations_stage,
                        wrangler_layer=self._wrangler_layer,
                        app=self._app,
                        org=self._org,
                        runtime=lmbda.Runtime.PYTHON_3_9,
                    )
                elif pipeline_type == CustomPipeline.PIPELINE_TYPE:
                    pipeline = CustomPipeline(
                        self,
                        construct_id=f"{team}-{pipeline_type}",
                        environment_id=self._environment_id,
                        resource_prefix=self._resource_prefix,
                        team=team,
                        foundations_stage=self._foundations_stage,
                        wrangler_layer=self._wrangler_layer,
                        app=self._app,
                        org=self._org,
                        runtime=lmbda.Runtime.PYTHON_3_9,
                    )
                else:
                    raise NotImplementedError(
                        f"Could not find a valid implementation for pipeline type: {pipeline_type}"
                    )
                pipelines[pipeline_name] = pipeline
            else:
                pipeline = pipelines[pipeline_name]

            # Register dataset to pipeline with concrete implementations
            dataset_name = f"{team}-{dataset}"
            if dataset_name not in dataset_names:
                dataset_names.add(dataset_name)
                pipeline.register_dataset(
                    dataset, config=customer_config.get("config", {})
                )
        
    def _create_wrangler_layer(self):
        wrangler_layer_version = lmbda.LayerVersion.from_layer_version_arn(
            self,
            "wrangler-layer",
            layer_version_arn=f"arn:aws:lambda:{self.region}:336392948345:layer:AWSDataWrangler-Python39:1",
        )

        return wrangler_layer_version

    def _create_data_lake_library_layer(self) -> lmbda.LayerVersion:
        data_lake_library_layer = lmbda.LayerVersion(
            self,
            "data-lake-library-layer",
            layer_version_name="data-lake-library",
            code=lmbda.Code.from_asset(
                os.path.join(
                    f"{Path(__file__).parents[1]}", "src/layers/data_lake_library"
                )
            ),
            compatible_runtimes=[lmbda.Runtime.PYTHON_3_9],
            description=f"{self._resource_prefix} Data Lake Library",
            license="Apache-2.0",
        )

        ssm.StringParameter(
            self,
            f"data-lake-library-layer-ssm",
            parameter_name=f"/SDLF/Layer/DataLakeLibrary",
            string_value=data_lake_library_layer.layer_version_arn,
        )
